backends/tfhe-cuda-backend/cuda/src/utils/helper_multi_gpu.cuh

```diff
@@ -0,0 +1,183 @@
+      cuda_memcpy_async_gpu_to_gpu(
+          d_dest, d_src, inputs_on_gpu * elements_per_input * sizeof(Torus),
+          streams[i], gpu_indexes[i]);
+
+    } else {
+      auto src_indexes = h_src_indexes + gpu_offset;
+
+      for (uint j = 0; j < inputs_on_gpu; j++) {
+        auto d_dest = dest[i] + j * elements_per_input;
+        auto d_src = src + src_indexes[j] * elements_per_input;
+
+        cuda_memcpy_async_gpu_to_gpu(d_dest, d_src,
+                                     elements_per_input * sizeof(Torus),
+                                     streams[i], gpu_indexes[i]);
+      }
+    }
+  }
+
+  if (sync_threads)
+    for (uint i = 0; i < active_gpu_count; i++)
+      cuda_synchronize_stream(streams[i], gpu_indexes[i]);
+}
+
+/// Copy data from multiple GPUs back to GPU 0 following the indexing given in
+/// dest_indexes
+/// The input indexing should be the trivial one
+template <typename Torus>
+void multi_gpu_gather_lwe(cudaStream_t *streams, uint32_t *gpu_indexes,
+                          uint32_t gpu_count, Torus *dest,
+                          const std::vector<Torus *> &src,
+                          Torus *h_dest_indexes, bool is_trivial_index,
+                          uint32_t num_inputs, uint32_t elements_per_input,
+                          bool sync_threads = true) {
+
+  auto active_gpu_count = get_active_gpu_count(num_inputs, gpu_count);
+
+  if (sync_threads)
+    cuda_synchronize_stream(streams[0], gpu_indexes[0]);
+
+  for (uint i = 0; i < active_gpu_count; i++) {
+    auto inputs_on_gpu = get_num_inputs_on_gpu(num_inputs, i, active_gpu_count);
+    auto gpu_offset = 0;
+    for (uint j = 0; j < i; j++) {
+      gpu_offset += get_num_inputs_on_gpu(num_inputs, j, active_gpu_count);
+    }
+
+    if (is_trivial_index) {
+      auto d_dest = dest + gpu_offset * elements_per_input;
+      auto d_src = src[i];
+
+      cuda_memcpy_async_gpu_to_gpu(
+          d_dest, d_src, inputs_on_gpu * elements_per_input * sizeof(Torus),
+          streams[i], gpu_indexes[i]);
+    } else {
+      auto dest_indexes = h_dest_indexes + gpu_offset;
+
+      for (uint j = 0; j < inputs_on_gpu; j++) {
+        auto d_dest = dest + dest_indexes[j] * elements_per_input;
+        auto d_src = src[i] + j * elements_per_input;
+
+        cuda_memcpy_async_gpu_to_gpu(d_dest, d_src,
+                                     elements_per_input * sizeof(Torus),
+                                     streams[i], gpu_indexes[i]);
+      }
+    }
+  }
+
+  if (sync_threads)
+    for (uint i = 0; i < active_gpu_count; i++)
+      cuda_synchronize_stream(streams[i], gpu_indexes[i]);
+}
+template <typename Torus>
+void multi_gpu_release_lwe(cudaStream_t *streams, uint32_t *gpu_indexes,
+                           std::vector<Torus *> &vec,
+                           bool sync_threads = true) {
+
+  for (uint i = 0; i < vec.size(); i++) {
+    cuda_drop_async(vec[i], streams[i], gpu_indexes[i]);
+    if (sync_threads)
```

Comment:
We probably want to launch all asynchronous memory release and then have a second loop synchronizing. Doing that we may have released some arrays before we get to the related synchronization call.

Of course, performance impact would not be huge but it's a small sand grain.

Reply:
I don't understand what you mean: here we release memory then synchronize, for each GPU. It seems to me this is equivalent to doing:
```
  for (uint i = 0; i < vec.size(); i++) {
      cuda_drop_async(vec[i], streams[i], gpu_indexes[i]);
  }
  if (sync_threads) {
      for (uint i = 0; i < vec.size(); i++) {
          cuda_synchronize_stream(streams[i], gpu_indexes[i]);
      }
  }
  vec.clear();
```
Am I wrong?

Reply:
Hmm ok I see what you mean. Indeed it should have a very small impact on performance, I'll make the change anyway.

Reply:
Actually, is it legitimate to call vec.clear() if we don't synchronize the threads? Isn't it going to remove some pointers before they're dropped?